# Classifying IMDB movie reviews

In [1]:
import turicreate as tc
import utils

In [2]:
movies = tc.SFrame('./IMDB_Dataset.csv')
movies

Finished parsing file /home/qyu2/SupervisedML/Assignment4/IMDB_Dataset.csv

Parsing completed. Parsed 100 lines in 0.361338 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[str,str]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 39624 lines. Lines per second: 59594.8

Finished parsing file /home/qyu2/SupervisedML/Assignment4/IMDB_Dataset.csv

Parsing completed. Parsed 50000 lines in 0.823681 secs.

Columns:
	review	str
	sentiment	str

Rows: 50000

Data:
+-------------------------------+-----------+
|             review            | sentiment |
+-------------------------------+-----------+
| One of the other reviewers... |  positive |
| A wonderful little product... |  positive |
| I thought this was a wonde... |  positive |
| Basically there's a family... |  negative |
| Petter Mattei's "Love in t... |  positive |
| Probably my all-time favor... |  positive |
| I sure would like to see a... |  positive |
| This show was an amazing, ... |  negative |
| Encouraged by the positive... |  negative |
| If you like original gut w... |  positive |
+-------------------------------+-----------+
[50000 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### 2a) There are 50,000 rows and 2 columns in the dataset, each row includes one item so there are 50,000 items are in the dataset. Each row includes the text of one review and the sentiment of that review (either positive or negative). The column for the text of review is named as review, and the column for the sentiment of review is named as sentiment. 

### 2b) The sentiment of review is labelled as either "positive" or "negative". 

In [3]:
movies['words'] = tc.text_analytics.count_words(movies['review'])
movies

Columns:
	review	str
	sentiment	str
	words	dict

Rows: 50000

Data:
+-------------------------------+-----------+-------------------------------+
|             review            | sentiment |             words             |
+-------------------------------+-----------+-------------------------------+
| One of the other reviewers... |  positive | {'darker': 1.0, 'touch': 1... |
| A wonderful little product... |  positive | {'done': 1.0, 'surface': 1... |
| I thought this was a wonde... |  positive | {'go': 1.0, 'superman': 1.... |
| Basically there's a family... |  negative | {'them': 1.0, 'ignore': 1.... |
| Petter Mattei's "Love in t... |  positive | {'work': 1.0, 'for': 1.0, ... |
| Probably my all-time favor... |  positive | {'for': 1.0, 'dozen': 1.0,... |
| I sure would like to see a... |  positive | {'do': 1.0, 'go': 1.0, 's'... |
| This show was an amazing, ... |  negative | {'awful': 1.0, 'just': 1.0... |
| Encouraged by the positive... |  negative | {'effort': 1.0, 'an': 1.0,... |
| If you like original gut w... |  positive | {'camp': 1.0, 'great': 1.0... |
+-------------------------------+-----------+-------------------------------+
[50000 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### 2c) The tc.text_analytics.count_words() function is a built-in Turi Create function from the text_analytics package. The function turns the text of a sentence into a dictionary with the word counts. The texts of the review are stored as a string, then the function counts the occurrences of the word for each review. Each word is stored as the key of the dictionary, and the occurrences of the word are the value in that dictionary. 

In [4]:
model = tc.logistic_classifier.create(movies, features=['words'], target='sentiment')

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 47500

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 101660

Number of coefficients      : 101661

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 2        | 1.000000  | 1.164903     | 0.917600          | 0.856800            |

| 1         | 4        | 1.000000  | 1.343996     | 0.936274          | 0.840000            |

| 2         | 9        | 0.630543  | 1.693933     | 0.966211          | 0.877600            |

| 3         | 11       | 0.630543  | 1.861909     | 0.975263          | 0.884800            |

| 4         | 13       | 0.630543  | 2.037939     | 0.983011          | 0.896800            |

| 9         | 18       | 1.000000  | 2.615833     | 0.999600          | 0.882400            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

### 2d) We created a words column to store the word counts into a dictionary. We process the review as the string, and each word will be a different feature. The number of words that appear in a review will be used to predict whether the review is positive or negative. The result is based on the number of positive words and the number of negative words that occur in a review. 

### 2e) From the table above, we can see that the iteration started from 0 and ended at 9, so the model iterates 10 times so there are 10 epochs of training are done.

### 2f) 5% of the data are used as the validation set. From the table above, noticed that the training accuracy for iteration 0, 1, 2, 3, 4, 9 are 0.917600, 0.936274, 0.966211, 0.975263, 0.983011, and 0.999600 respectively. The validation accuracy for iteration 0, 1, 2, 3, 4, 9 are 0.856800, 0.840000, 0.877600, 0.884800, 0.896800, and 0.882400 respectively. The model ended with a training accuracy of 0.999600 and a validation accuracy of 0.882400. The model is able to perform with 88.24% accuracy on new data. The training accuracy is higher.  It is expected because only 5% of data are used as the validation set, which the dataset set is used to check how well the model is able to make predictions based on data it hasn't seen before. The model is already familiar with the training dataset but the validation dataset is new to the model; therefore, the training accuracy is expected to be high than the training accuracy.  

In [5]:
model

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 101661
Number of examples             : 47500
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 101660

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 0.01

Training Summary
----------------
Solver                         : lbfgs
Solver iterations              : 10
Solver status                  : Completed (Iteration limit reached).
Training time (sec)            : 1.1867

Settings
--------
Log-likelihood                 : 472.2715

Highest Positive Coefficients
-----------------------------
words[sendak]                  : 44.8536
words[gic]                     : 44.8536
words[widdecombe]              : 32.0042
words[distrusting]             : 32.0042
words[aleination]              : 28.9013

Lowest Negative Coefficients
----------------------------
words[drosselmeier]           

### 2g) There are 101,661 of coefficients are in the model.

### 2h) The weight to the L1 penalty for regularization is 0 and the weight to the L2 penalty for regulariztion is 0.01. 

In [6]:
weights = model.coefficients
weights

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: 101661

Data:
+-------------+---------+----------+-----------------------+--------+
|     name    |  index  |  class   |         value         | stderr |
+-------------+---------+----------+-----------------------+--------+
| (intercept) |   None  | positive |  0.11321421383120678  |  None  |
|    words    |  darker | positive |  0.48004309270921713  |  None  |
|    words    |  touch  | positive |  0.20644753896367019  |  None  |
|    words    |  thats  | positive |  -0.3977478976394822  |  None  |
|    words    |   your  | positive | -0.020570217177958334 |  None  |
|    words    | viewing | positive |  0.12541121536462171  |  None  |
|    words    |  their  | positive |  0.016792942314132323 |  None  |
|    words    |   into  | positive | -0.004248681285055308 |  None  |
|    words    |  turned | positive |  -0.24462918867409988 |  None  |
|    words    |  being  | positive | -0.017940915871508683 |  None  |
+-------------+---------+----------+-----------------------+--------+
[101661 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [7]:
weights.sort('value')

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: 101661

Data:
+-------+----------------+----------+---------------------+--------+
|  name |     index      |  class   |        value        | stderr |
+-------+----------------+----------+---------------------+--------+
| words |  drosselmeier  | positive |  -70.90827141379647 |  None  |
| words |      eta       | positive |  -35.91368312140312 |  None  |
| words |    nanites     | positive |  -24.83955701603244 |  None  |
| words | choreographic  | positive |  -21.00350980517402 |  None  |
| words |    newlwed     | positive | -13.425744255431141 |  None  |
| words |     sierre     | positive | -13.379433032016859 |  None  |
| words |    poolguy     | positive | -12.984192095535354 |  None  |
| words |      bsed      | positive | -12.930967704749738 |  None  |
| words |  nonsenseful   | positive | -12.892688768200786 |  None  |
| words | unfortuntately | positive | -12.863293183865931 |  None  |
+-------+----------------+----------+---------------------+--------+
[101661 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [8]:
weights.sort('value', ascending=False)

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: 101661

Data:
+-------+---------------+----------+-------------------+--------+
|  name |     index     |  class   |       value       | stderr |
+-------+---------------+----------+-------------------+--------+
| words |     sendak    | positive |  44.8535807070221 |  None  |
| words |      gic      | positive |  44.8535807070221 |  None  |
| words |  distrusting  | positive | 32.00415207623371 |  None  |
| words |   widdecombe  | positive | 32.00415207623371 |  None  |
| words |   insititue   | positive | 28.90125182206651 |  None  |
| words |    artisty    | positive | 28.90125182206651 |  None  |
| words |   embroider   | positive | 28.90125182206651 |  None  |
| words |   aleination  | positive | 28.90125182206651 |  None  |
| words |     blain     | positive | 25.14985973615453 |  None  |
| words | kaleidoscopic | positive | 25.14985973615453 |  None  |
+-------+---------------+----------+-------------------+--------+
[101661 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [28]:
print("2i) The top 10 words that contributed to positive reviews are", weights.sort('value', ascending=False)["index"][:10])

2i) The top 10 words that contributed to positive reviews are ['sendak', 'gic', 'distrusting', 'widdecombe', 'insititue', 'artisty', 'embroider', 'aleination', 'blain', 'kaleidoscopic']


### 2i) The top 10 words that contributed to positive reviews are 'sendak', 'gic', 'distrusting', 'widdecombe', 'insititue', 'artisty', 'embroider', 'aleination', 'blain', 'kaleidoscopic'.

In [30]:
print("2j) The top 10 words that contributed to negative reviews are", weights.sort('value')["index"][:10])

2j) The top 10 words that contributed to negative reviews are ['drosselmeier', 'eta', 'nanites', 'choreographic', 'newlwed', 'sierre', 'poolguy', 'bsed', 'nonsenseful', 'unfortuntately']


### 2j) 2j) The top 10 words that contributed to negative reviews are 'drosselmeier', 'eta', 'nanites', 'choreographic', 'newlwed', 'sierre', 'poolguy', 'bsed', 'nonsenseful', 'unfortuntately'.

In [9]:
weights[weights['index']=='wonderful']

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: Unknown

Data:
+-------+-----------+----------+--------------------+--------+
|  name |   index   |  class   |       value        | stderr |
+-------+-----------+----------+--------------------+--------+
| words | wonderful | positive | 0.9969894541128703 |  None  |
+-------+-----------+----------+--------------------+--------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [10]:
weights[weights['index']=='horrible']

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: Unknown

Data:
+-------+----------+----------+---------------------+--------+
|  name |  index   |  class   |        value        | stderr |
+-------+----------+----------+---------------------+--------+
| words | horrible | positive | -0.9998770751123641 |  None  |
+-------+----------+----------+---------------------+--------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [48]:
weights[weights['index']=='the']

Columns:
	name	str
	index	str
	class	str
	value	float
	stderr	float

Rows: Unknown

Data:
+-------+-------+----------+-----------------------+--------+
|  name | index |  class   |         value         | stderr |
+-------+-------+----------+-----------------------+--------+
| words |  the  | positive | 0.0004862836915785153 |  None  |
+-------+-------+----------+-----------------------+--------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

### 2k) The weight is given to the word 'the' is 0.0004862836915785153. No, this is not a sensible value, because the word 'the' has not much meaning to either the positive or negative and it occurs in abundance in the reviews. 

### 2l) To improve this model, I will try to remove the stop words such as 'the', 'a', ' is', 'if' and so on. There are 101,661 coefficients are in the model. Removing these words can reduce the number of coefficients. Reducing the number of non-important features can help the model focus more on the important features. These words occur in abundance and they are usually not important or provide little information that can be used in sentiment analysis. We want to remove these words as much as we can to give the model more focus on the important information for prediction. These words have no meaning or value to contribute with either positive or negative sentiment. Even these words had a very small amount of weight to the model; they still slightly affect the output of the algorithm. Another way to improve the model is by normalizing the terms, some words have the same meaning but display in different tense or grammar purposes (noun, verb, adjective, and so on). For example, "mention" and "mentioned", "watch" and "watching", "movie" and "movies" have the same effect for sentiment analysis. 

In [12]:
movies['predictions'] = model.predict(movies, output_type='probability')

In [13]:
movies.sort('predictions', ascending=False)[0]

{'review': "The effects of job related stress and the pressures born of a moral dilemma that pits conscience against the obligations of a family business (albeit a unique one) all brought to a head by-- or perhaps the catalyst of-- a midlife crisis, are examined in the dark and absorbing drama, `Panic,' written and directed by Henry Bromell, and starring William H. Macy and Donald Sutherland. It's a telling look at how indecision and denial can bring about the internal strife and misery that ultimately leads to apathy and that moment of truth when the conflict must, of necessity, at last be resolved.<br /><br />\tAlex (Macy) is tired; he has a loving wife, Martha (Tracey Ullman), a precocious six-year-old son, Sammy (David Dorfman), a mail order business he runs out of the house, as well as his main source of income, the `family' business he shares with his father, Michael (Sutherland), and his mother, Deidre (Barbara Bain). But he's empty; years of plying this particular trade have le

In [14]:
movies.sort('predictions', ascending=True)[0]

{'review': "The Nutcracker has always been a somewhat problematic ballet. It bears little resemblance to ETA Hoffman's original story on which it is based.<br /><br />In the ballet, the story is essentially over by the second-half when Clara (or Marie in this version) travels to the Kingdom of Sweets to watch a series of character dances.<br /><br />There's an infinite variety of stage productions that re-interpret the story in myriad ways (not always successfully) to compensate for the ballet's weak libretto.<br /><br />Balanchine's version doesn't really have any sense of drama or story at all (despite the fact that there is plenty of drama and mystery in Tchaikovsky's wonderful first-act music). The result is a completely forgettable first-half Christmas party where hardly anything happens and where even the dancing (the little that there is of it) isn't particularly memorable.<br /><br />The pantomime over-acting, particularly of Drosselmeier, which might look passable on the stage